### PR TITLE
Enhance rendezvous docs and fix race condition

### PR DESCRIPTION
**Pull request by ghpvnist**
The PR removes the use of `shared_ptr`s as it complicates the logic of `rendezvous`. Instead of relying on the use count of `shared_ptr`, I've added a variable to track the state separately instead. Now, there is no need to worry about the scope of the `shared_ptr`s and cleaning up resource before exiting.

Thanks @mlevesquedion for sharing code snippets offline to help simplify this!

### CODE REVIEW

**Comment by mlevesquedion** [comment]
Looks like I hit this race condition in #1960: https://github.com/openxla/stablehlo/actions/runs/7675148587/job/20920959254?pr=1960

**Comment by mlevesquedion** [comment]
Hit it again: https://github.com/openxla/stablehlo/actions/runs/7703298470/job/20993390030#step:6:437

I'm really unlucky :upside_down_face:

**Comment by GleasonK** [comment]
Based on Michaels comments I'm not sure if this fixes the issue?

**Review by GleasonK — APPROVED** [review]

**Comment by GleasonK** [comment]
Ah - or did I misunderstand, perhaps Michael is hitting this because this fix isnt in.

**Comment by ghpvnist** [comment]
Yes, it's the latter. Though, I'm thinking about simplifying this so I'll hold off on merging just yet.

File: stablehlo/reference/ProcessGrid.h [hunk 1]
```diff
@@ -42,15 +67,20 @@ namespace detail {
 /// Once all processes have added their data, the data in `values` is moved to
 /// `result` that multiple processes can concurrently read from.
 struct RendezvousState {
+ public:
```
Comment (mlevesquedion) [inline stablehlo/reference/ProcessGrid.h:70]: `public` is implied by `struct`

File: stablehlo/reference/ProcessGrid.cpp [hunk 2]
```diff
@@ -220,72 +220,32 @@ SmallVector<Tensor> ProcessGrid::recv(ChannelId channelId,
...
-    // The last process to contribute takes the result from the state to allow
-    // the process that contributed last to exit the function.
-    auto result = std::move(state.result);
-    channelConditions_[channelKey].notify_one();
-    return result;
+  state.useCount++;
+
+  // After each process contributes, wait for the last process to notify.
+  if (state.values.size() < processGroup.size()) {
+    channelConditions_[channelKey].wait(lock);
```
Comment (mlevesquedion) [inline stablehlo/reference/ProcessGrid.cpp:240]: Should we have a limit on the wait time? Used to be 3 seconds.

File: stablehlo/reference/ProcessGrid.h (hunk 1 above)
Comment (ghpvnist) [reply stablehlo/reference/ProcessGrid.h:70]: Thanks, I missed updating this while debugging :)

File: stablehlo/reference/ProcessGrid.cpp (hunk 2 above)
Comment (mlevesquedion) [reply stablehlo/reference/ProcessGrid.cpp:240]: (The 3 seconds is documented in `ProcessGrid.h` too, so consider updating that if you decide to remove the wait limit.)

File: stablehlo/reference/ProcessGrid.cpp [hunk 3]
```diff
@@ -220,72 +220,32 @@ SmallVector<Tensor> ProcessGrid::recv(ChannelId channelId,
...
-    if (state.result.use_count() > static_cast<int64_t>(processGroup.size()))
-      llvm::report_fatal_error(
-          "Each process should have only one shared access to the result.");
-
-    // The last process to contribute takes the result from the state to allow
-    // the process that contributed last to exit the function.
-    auto result = std::move(state.result);
-    channelConditions_[channelKey].notify_one();
-    return result;
+  state.useCount++;
```
Comment (mlevesquedion) [inline stablehlo/reference/ProcessGrid.cpp:236]: Doesn't `state.result` already have a `use_count` helper? Would it be simpler to use that instead?

**Review by GleasonK — CHANGES_REQUESTED** [review]
Going to re-review with code changes. Adding blocking review until then.

File: stablehlo/reference/ProcessGrid.cpp (hunk 3 above)
Comment (ghpvnist) [reply stablehlo/reference/ProcessGrid.cpp:236]: With the refactoring, we don't have shared_ptrs anymore. `state.result` is a map (size only increasing) so we can't rely on that.

Comment (mlevesquedion) [reply stablehlo/reference/ProcessGrid.cpp:236]: Ah yep my bad, didn't realize this was wrapped in a `shared_ptr` at first. (Also I didn't mean to resolve this without commenting. I really wish the GH UI would either submit the comment or at least warn when you have a pending comment and you hit "resolve".

**Review by mlevesquedion — APPROVED** [review]
LGTM

File: stablehlo/reference/ProcessGrid.cpp [hunk 4]
```diff
@@ -220,72 +220,35 @@ SmallVector<Tensor> ProcessGrid::recv(ChannelId channelId,
...
-
-    // The last process to contribute takes the result from the state to allow
-    // the process that contributed last to exit the function.
-    auto result = std::move(state.result);
-    channelConditions_[channelKey].notify_one();
-    return result;
+            lock, std::chrono::seconds(3),
+            [&] { return state.values.size() == processGroup.size(); }))
+      llvm::report_fatal_error("rendezvous timed out");
+  } else {
```
Comment (GleasonK) [inline stablehlo/reference/ProcessGrid.cpp:244]: If stmt with an exit in the body, do we need the `else`?

Comment (GleasonK) [reply stablehlo/reference/ProcessGrid.cpp:244]: Ah. Resolving. This is a nested if.

**Review by GleasonK — APPROVED** [review]

**Pull request merged**